sematic/plugins/abstract_metrics_storage.py

```diff
@@ -0,0 +1,202 @@
+    )
+    ```
+    """
+
+    metric_name: str
+    metric_type: Optional[str] = None
+    series: List[Tuple[float, Tuple[Union[str, int], ...]]] = field(
+        default_factory=list
+    )
+    columns: List[str] = field(default_factory=list)
+
+
+class GroupBy(enum.Enum):
+    """
+    Options to aggregate metrics over.
+    """
+
+    run_id = "run_id"
+    calculator_path = "calculator_path"
+    root_id = "root_id"
+    root_calculator_path = "root_calculator_path"
+
+
+class NoMetricError(Exception):
+    def __init__(self, metric_name: str, plugin_name: str):
+        super().__init__(f"No metric named {metric_name} was found in {plugin_name}")
+
+
+class AbstractMetricsStorage(abc.ABC):
+    """
+    Abstract base class for metrics storage plug-ins.
+    """
+
+    @abc.abstractmethod
+    def store_metrics(self, metric_points: Sequence[MetricPoint]) -> None:
+        """
+        Persists a list of metric points.
+
+        Parameters
+        ----------
+        metric_points: Iterable[MetricPoint]
+            List of metric points to persist.
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_metrics(self, labels: MetricsLabels) -> Tuple[str, ...]:
+        """
+        Gets the list of unique metric names for the given set of labels.
+
+        Parameters
+        ----------
+        labels: MetricsLabels
+            Labels by which to filter metric names.
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_aggregated_metrics(
+        self,
+        filter: MetricsFilter,
+        group_by: Sequence[GroupBy],
+        rollup: Union[int, Literal["auto"], None],
+    ) -> MetricSeries:
+        """
+        Returns a metric aggregation according to the provided filter and group
+        bys.
+
+        Parameters
+        ----------
+        filter: MetricsFilter
+            Criteria to filter metrics values.
+        group_by: Iterable[GroupBy]
+            List of dimensions to aggregate metrics over.
+        rollup: Union[int, Literal["auto"], None]
+            How to aggregate metrics over time. `None` means no aggregation,
+            returns a single scalar per combination of `group_by` dimensions.
+            `"auto"` means the optimal aggregation interval will be found to
+            have a maximum of 300 series points. Passing an int sets the
+            interval size in seconds. If this interval is capped to yield no
+            more than 300 buckets.
+        Examples
+        --------
+        All time run count ``` MetricsStorage().get_aggregated_metrics(
+            filter=MetricsFilter(
+                name="sematic.func_run_count",
+                from_time=datetime.fromtimestamp(0), to_time=datetime.utcnow(),
+                labels={}
+            ), group_by=[], rollup=None,
+        )
+        ```
+
+        Success rate for Function `path.to.foo` by date over the last 30 days
+        ``` MetricsStorage().get_aggregated_metrics(
```

Comment:
```suggestion
        Success rate for Function `path.to.foo` by date over the last 30 days:
        ```
        MetricsStorage().get_aggregated_metrics(
```